libsignal-service-dotnet/websocket/WebSocketWrapper.cs

```diff
@@ -29,36 +29,96 @@ public class WebSocketWrapper
 
         public WebSocketWrapper(string uri, CancellationToken token)
         {
-            _ws = new ClientWebSocket();
-            _ws.Options.KeepAliveInterval = TimeSpan.FromSeconds(20);
+            CreateSocket();
             _uri = new Uri(uri);
             Token = token;
         }
+        private void CreateSocket()
+        {
+            _ws = new ClientWebSocket();
+            _ws.Options.KeepAliveInterval = TimeSpan.FromSeconds(20);
+        }
 
         public void HandleOutgoingWS()
         {
             Debug.WriteLine(TAG + "HandleOutgoingWS started");
+            byte[] buf = null;
             while (!Token.IsCancellationRequested)
             {
+                
                 try
                 {
-                    var buf = OutgoingQueue.Take(Token);
+                    if (buf == null)
+                        buf = OutgoingQueue.Take(Token);
+                    ReconnectTask?.Wait();
                     _ws.SendAsync(new ArraySegment<byte>(buf, 0, buf.Length), WebSocketMessageType.Binary, true, Token).Wait();
+                    buf = null; //set to null so we do not retry the same block
                 }
                 catch (TaskCanceledException e)
                 {
+                    if (!Token.IsCancellationRequested)
+                        Reconnect();
                     Debug.WriteLine(TAG + "HandleOutgoingWS shutting down");
                 }
                 catch (Exception e)
                 {
                     Debug.WriteLine(TAG + "WS SendAsync failed: " + e.Message);
-                    //TODO reconnect
+                    Reconnect();
                 }
             }
             //TODO dispose
             Debug.WriteLine(TAG + "HandleOutgoingWS finished");
         }
+        private Task ReconnectTask;
+        public void Reconnect()
+        {
+            if (Token.IsCancellationRequested)
+                return;
+            if (ReconnectTask != null)
+            {
+                ReconnectTask?.Wait();//race condition handling
```

Comment:
Yeah I was aware that a read/write could cause a race condition causing two reconnects to occur at the exact same time.  Technically this would not do any harm (just two connects instead of one) but a lock could avoid it.    I am not 100% confident that the socket state will always be correct so rather than using that to see if a reconnect has happened I think I will grab a copy of the WS at the start of the function and at the start of the lock if the WS has changed can safely assume a reconnect has happened.

If this sounds good let me know and will push it up.